File: app/src/main/java/me/ykrank/s1next/widget/RangeInputFilter.java
package me.ykrank.s1next.widget;

import android.text.InputFilter;
import android.text.Spanned;

import com.google.common.collect.Range;

import org.apache.commons.lang3.StringUtils;

/**
 * Constrains input from entering a value which is out of the range.
 * <p>
 * Forked from http://stackoverflow.com/q/14212518
 */
public final class RangeInputFilter implements InputFilter {

    private final Range<Integer> mRange;

    public RangeInputFilter(int lower, int upper) {
        this.mRange = Range.closed(lower, upper);
    }

    @Override
    public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
        String value = dest.subSequence(0, dstart).toString()
                + source.subSequence(start, end)
                + dest.subSequence(dend, dest.length());
        try {
            if (mRange.contains(Integer.valueOf(value))) {
                return null;
            }
        } catch (NumberFormatException ignored) {

        }

        return StringUtils.EMPTY;
    }
}

File: library/src/main/java/com/github/ykrank/androidtools/widget/RangeInputFilter.kt
package com.github.ykrank.androidtools.widget

import android.text.InputFilter
import android.text.Spanned
import android.util.Range

/**
 * Constrains input from entering a value which is out of the range.
 *
 *
 * Forked from http://stackoverflow.com/q/14212518
 */
class RangeInputFilter(lower: Int, upper: Int) : InputFilter {
    private val mRange: Range<Int>

    init {
        mRange = Range.create(lower, upper)
    }

    override fun filter(
        source: CharSequence,
        start: Int,
        end: Int,
        dest: Spanned,
        dstart: Int,
        dend: Int
    ): CharSequence? {
        val value = (dest.subSequence(0, dstart).toString()
                + source.subSequence(start, end)
                + dest.subSequence(dend, dest.length))
        try {
            if (mRange.contains(value.toInt())) {
                return null
            }
        } catch (ignored: NumberFormatException) {
        }
        return ""
    }
}
